### PR TITLE
Data cleaning script

### PR DESCRIPTION
**Pull request by tk755**
To run script:
```
gcloud dataproc jobs submit pyspark --cluster ${CLUSTER_NAME} \
    --jars gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar \
    --driver-log-levels root=FATAL \
    clean.py -- ${PROJECT_ID} ${BUCKET_NAME}
```

### CODE REVIEW

File: data-science-onramp/data-cleaning/clean.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            return 'Male'
+        elif s.lower().startswith('f'):
+            return 'Female'
+        else:
+            return None
+
+def angle(a):
+    '''Converts DMS notation to angles.'''
+    if a:
+        dms = re.match('(-?\d*).(-?\d*)\'(-?\d*)"', a)
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: Does this check if there is a degrees symbol? You may want to check for it using the unicode value: "\u00B0"

File: data-science-onramp/data-cleaning/clean.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    for name, obj in multi_udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+
+    if upload:
+        backfill(df)
+    else:
+        df.show(n=100)
+
+if __name__ == '__main__':
+    main()
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: Add a newline here

File: data-science-onramp/data-cleaning/clean.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+
+def main():
+    # Create a SparkSession under the name 'clean'. Viewable via the Spark UI
+    spark = SparkSession.builder.appName('clean').getOrCreate()
+
+    # Whether to backfill data to GCS bucket
+    upload = True
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: It might be better to get the upload value from the arguments. For example, if somebody adds a --test flag, we can make upload = false. We used a similar approach like this in setup.py

File: data-science-onramp/data-cleaning/clean.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        }
+    }
+
+    for name, obj in multi_udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+
+    if upload:
+        backfill(df)
+    else:
+        df.show(n=100)
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: I think you should get a sample of the dataset and then show it like this: `new_df.sample(False, 0.0001).show(n=100)` This is because when I checked the job output, all of the genders were female for some reason? Not showing a random sample might cause problems when testing.

File: data-science-onramp/data-cleaning/clean.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            'params': ('tripduration', 'starttime', 'stoptime')
+        }
+    }
+
+    for name, obj in multi_udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+
+    if upload:
+        backfill(df)
+    else:
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: I think you should print all the time, even when you are uploading because you will need to check the output during testing

File: data-science-onramp/data-cleaning/clean.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+from pyspark.sql.types import StringType, IntegerType, FloatType
+
+
+PROJECT_ID = sys.argv[1]
+BUCKET_NAME = sys.argv[2]
+TABLE = f'{PROJECT_ID}.new_york_citibike_trips.RAW_DATA'
+
+def trip_duration(duration):
+    '''Convert trip duration to seconds.'''
+    if duration:
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: nit: I would recommend switching from this:
```
if exists:
    <logic>
```
to this:
```
if not exist:
    return None

<logic>
```

This makes the code look a bit cleaner via less indentation

File: data-science-onramp/data-cleaning/clean.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        else:
+            if number < 0:
+                return None
+            else:
+                return int(number)
+
+def station_name(name):
+    '''Replaces '/' with '&'.'''
+    if name:
+        return name.replace('/', '&')
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: suggestion: `return name.replace('/', '&') if name else None`

File: data-science-onramp/data-cleaning/clean.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+
+def main():
+    # Create a SparkSession under the name 'clean'. Viewable via the Spark UI
+    spark = SparkSession.builder.appName('clean').getOrCreate()
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: Maybe the app name should be a little more descriptive like Data_Processing?

File: data-science-onramp/data-cleaning/clean.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
+import os
+import sys
+import re
+import datetime
+
+from py4j.protocol import Py4JJavaError
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import UserDefinedFunction, lit
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: Are you using lit?

File: data-science-onramp/data-cleaning/clean.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        'start_station_longitude': (angle, FloatType()),
+        'end_station_latitude': (angle, FloatType()),
+        'end_station_longitude': (angle, FloatType())
+    }
+
+    for name, udf in udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+
+    # Multiple-parameter column transformations
+    multi_udfs = {
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py:164]: Another multiple-parameter function Diego and I discussed when thinking of ways to dirty the script was for station ID and location/name. For example, if a station location or name is null, we were thinking of finding another row with the same station ID and getting the correct information from there if it's not null. There should not be any null station ID's because we did not touch that column

File: data-science-onramp/data-cleaning/clean.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    '''Converts user type to 'Subscriber' or 'Customer'.'''
+    if user:
+        if user.lower().startswith('sub'):
+            return 'Subscriber'
+        elif user.lower().startswith('cust'):
+            return 'Customer'
+        else:
+            return None
+
+def gender(s):
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: "s" is a kind of vague variable name

File: data-science-onramp/data-cleaning/clean.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
+import os
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: Are you using os?

File: data-science-onramp/data-cleaning/clean.py (hunk 1 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: This checks for any single character. I was going to use the unicode value but I couldn't get it working well using re. I thought this was a fair compromise because there are no other instances in the dataset that this particular regex would match.

File: data-science-onramp/data-cleaning/clean.py (hunk 11 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: I agree. Think I copied it from setup.py but I will change it

File: data-science-onramp/data-cleaning/clean.py (hunk 8 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: What do you think about 'data_cleaning'? I'm just hesitant to use the word 'processing' because I'm not sure if feature extraction falls under it as well.

File: data-science-onramp/data-cleaning/clean.py (hunk 11 above)
Comment (vuppalli) [reply data-science-onramp/data-cleaning/clean.py]: Haha, I will change it in setup.py as well!

File: data-science-onramp/data-cleaning/clean.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            return int(number)
+        elif 'm' in duration:
+            return int(number * 60)
+        elif 'h' in duration:
+            return int(number * 60 * 60)
+        else:
+            if number < 0:
+                return None
+            else:
+                return int(number)
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: You can remove a few conditionals by switching to something similar to the following:

```
# a negative duration, regardless, should probably be cleaned out
if not number or number < 0: 
    return None

number = float(number[0])

if 'm' in duration:
    number *= 60
elif 'h' in duration:
    number *= 60 * 60

return int(number)

```

Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean.py]: May want to use a more explicit variable name than `number` as well

File: data-science-onramp/data-cleaning/clean.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+def station_name(name):
+    '''Replaces '/' with '&'.'''
+    if name:
+        return name.replace('/', '&')
+
+def user_type(user):
+    '''Converts user type to 'Subscriber' or 'Customer'.'''
+    if user:
+        if user.lower().startswith('sub'):
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: similar to above
```
if not user: 
    return None
```

File: data-science-onramp/data-cleaning/clean.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+def gender(s):
+    '''Converts gender to 'Male' or 'Female' or '''
+    if s:
+        if s.lower().startswith('m'):
+            return 'Male'
+        elif s.lower().startswith('f'):
+            return 'Female'
+        else:
+            return None
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: similar to above:
```
if not s:
    return None
```

File: data-science-onramp/data-cleaning/clean.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: I would add a docstring here just explaining what it's doing.

**Review by bradmiro — CHANGES_REQUESTED** [review]

File: data-science-onramp/data-cleaning/clean.py [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+    '''Calculates duration from start and end time if null.'''
+    return compute_time(duration, start, end)[0]
+
+def compute_start(duration, start, end):
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: This is a little hard to follow, but I can't honestly think of a much better way to do this. We may need tweak it once we incorporate it into the docs, but for now I think it's fine.

**Comment by bradmiro** [comment]
You may want to change the naming convention of your UDFs from `<column_name>` to `column_name_udf`. This will make it a bit easier to follow what's a UDF vs a helper function as well as make changing some of the variable names from a letter to the column name. For instance, you can replace `def angle(a):` with `def angle_udf(angle):`

File: data-science-onramp/data-cleaning/clean.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+def backfill(df):
+    '''Writes a dataframe to a GCP bucket.'''
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+    return path
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: I don't think you need to return the path, but you should add a print statement here stating the load is done.

File: data-science-onramp/data-cleaning/clean.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+
+def compute_start(duration, start, end):
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
```
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: Change to `upload_to_gcs` perhaps?

File: data-science-onramp/data-cleaning/clean.py [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+TABLE = f'{PROJECT_ID}.new_york_citibike_trips.RAW_DATA'
+
+def trip_duration(duration):
+    '''Convert trip duration to seconds.'''
+    if not duration:
+        return None
+    
+    time = re.match('\d*.\d', duration)
+
+    if not time:
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.py:25]: Should this error out? Or provide any kind of logging info? Or is it appropriate to just ignore? (I'm asking not to be passive aggressive - I'm asking so i can make sure I genuinely understand why things are happening the way they are)

File: data-science-onramp/data-cleaning/clean.py (hunk 17 above)
Comment (leahecole) [reply data-science-onramp/data-cleaning/clean.py]: I think more comments might help make it understandable

**Review by leahecole — CHANGES_REQUESTED** [review]
My big meta question is if this is being run in a cluster, what are the implications of returning None throughout, and would it improve developer experience to also include more verbose print statements and/or to catch for more errors?

File: data-science-onramp/data-cleaning/clean.py (hunk 20 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py:25]: This case should not happen often. The format of the values in this column are either floats or ints, optionally followed by one of 's', 'm', or 'h'. This regex extracts the numerical part of the value, and then transforms it depending on the unit. It will fail to match (i.e. `not time` will be true) if there is no numerical value in the column, in which case it cannot be transformed (if there is no numerical value, then the value is not useful and should be dropped). To my knowledge, the dirty dataset doesn't contain anything that will cause this to occur, but the following lines will cause runtime errors if it does occur, so it is there as a preventative measure.

Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py:25]: And to answer your question of returning None throughout - that only happens when some data is encountered for which we don't know how to transform/clean (e.g. a negative time). For these cases, we just want to eliminate the value altogether. Maybe there is value in printing something when this happens. But my thought is that generally this is what we want, and also this will happen so often for so many nonsensical values that if there was something of value, it would be difficult to find. What do you think?

File: data-science-onramp/data-cleaning/clean.sh [hunk 21]
```diff
@@ -0,0 +1,12 @@
+# Submit a PySpark job via the Cloud Dataproc Jobs API
+# Requires having PROJECT_ID, CLUSTER_NAME and BUCKET_NAME set as 
+# environment variables
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.sh:3]: Also requires `CLOUDSDK_DATAPROC_REGION` environment variable, or for `gcloud config set dataproc/region VALUE`

Comment (leahecole) [reply data-science-onramp/data-cleaning/clean.sh:3]: Double check - should this value correspond to the region where our cluster is located? (the one specified in `CLUSTER_NAME`)

File: data-science-onramp/data-cleaning/clean.sh [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+
+export CLUSTER_NAME=data-cleaning
+export PROJECT_ID=data-science-onramp
+export BUCKET_NAME=citibikevd
+
+gcloud dataproc jobs submit pyspark --cluster ${CLUSTER_NAME} \
+    --jars gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar \
+    --driver-log-levels root=FATAL \
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.sh:11]: What does this flag do, and why do we have it set to the value it is set to?

File: data-science-onramp/data-cleaning/clean.py (hunk 20 above)
Comment (leahecole) [reply data-science-onramp/data-cleaning/clean.py:25]: I'm happy with this solution as long as we add a function level comment explaining what gets returned when it's successful and what gets returned when it's not

File: data-science-onramp/data-cleaning/clean.py [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+    if not angle:
+        return None
+    
+    dms = re.match('(-?\d*).(-?\d*)\'(-?\d*)"', angle)
+    if dms:
+        return int(dms[1]) + int(dms[2])/60 + int(dms[3])/(60 * 60)
+    
+    try:
+        return float(angle)
+    except ValueError:
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.py]: Per our conversation, this probably needs to be expanded to catch other cases that can be converted

If you do use an error, add a short comment explaining why `None` is returned, or add logging.

File: data-science-onramp/data-cleaning/clean.py [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+if __name__ == '__main__':
+    # Create a SparkSession, viewable via the Spark UI
+    spark = SparkSession.builder.appName('data_cleaning').getOrCreate()
+
+    # Load data into dataframe if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f'{TABLE} does not exist.')
+        raise
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.py]: Will there be a mess leftover when this terminates? Double check this

File: data-science-onramp/data-cleaning/clean.py [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    for name, obj in multi_udfs.items():
+        df = df.withColumn(name, obj['udf'](*obj['params']))
+
+    # Display sample of 100 rows
+    df.sample(False, 0.001).show(n=100)
+
+    # Write results to GCS
+    if '--test' in sys.argv:
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean.py]: Should we perhaps call this `--dry-run`

File: data-science-onramp/data-cleaning/clean_test.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+import re
+import uuid
+
+from google.cloud import dataproc_v1 as dataproc
+from google.cloud import storage
+import pytest
+
+# Set global variables
+PROJECT = os.environ['GCLOUD_PROJECT']
+REGION = "us-east4"
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py]: What is this region for? Is it for dataproc? A different resource?  Plz make the variable name reflect what resource it represents

File: data-science-onramp/data-cleaning/clean_test.py [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+    'config': {
+        'gce_cluster_config': {
+            'zone_uri': '',
+            "metadata": {
+                "PIP_PACKAGES": "google-cloud-storage"
+            },
+        },
+        'master_config': {
+            'num_instances': 1,
+            'machine_type_uri': 'n1-standard-8'
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py]: Should machine type and num instances also be global variables?

Comment (leahecole) [reply data-science-onramp/data-cleaning/clean_test.py]: @bradmiro

File: data-science-onramp/data-cleaning/clean_test.py [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+        },
+        'worker_config': {
+            'num_instances': 6,
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py]: same here wrt num_instances and machine type - variable? @bradmiro

File: data-science-onramp/data-cleaning/clean_test.py [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+            'machine_type_uri': 'n1-standard-8'
+        },
+        "initialization_actions": [
+            {
+                "executable_file": ("gs://dataproc-initialization-actions/"
+                                    "python/pip-install.sh"),
+            }
+        ],
+        "software_config": {
+            "image_version": "1.5.4-debian10",
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py]: Will this image version change often? Should it potentially be an environment variable? @bradmiro

File: data-science-onramp/data-cleaning/clean_test.py [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+            "optional_components": [
+                "ANACONDA"
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py]: I think I asked this for data ingestion too, why do we have to have this optional component?

File: data-science-onramp/data-cleaning/clean_test.py [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+    yield
+
+    # Delete GCS bucket
+    bucket = storage_client.get_bucket(BUCKET_NAME)
+    bucket.delete(force=True)
+
+def is_in_table(value, out):
+    return re.search(f"\\| *{value} *\\|", out)
+
+def get_blob_from_path(path):
```
Comment (leahecole) [inline data-science-onramp/data-cleaning/clean_test.py:82]: I feel like I asked this last time - sorry for forgetting - this is a dataproc specific thing, right? We can't pass this as an env var because Dataproc makes it along the way? @bradmiro

**Review by leahecole — CHANGES_REQUESTED** [review]
Make sure to fix lint things by
1. copy `noxfile-template.py` from `python-docs-samples/` to your folder and name it `noxfile.py`
2. Run `nox -s lint`
3. Fix what it tells you to fix

You can also use [`black`](https://pypi.org/project/black/) which will fix a lot of it

File: data-science-onramp/data-cleaning/clean_test.py [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+        'worker_config': {
+            'num_instances': 6,
+            'machine_type_uri': 'n1-standard-8'
+        },
+        "initialization_actions": [
+            {
+                "executable_file": ("gs://dataproc-initialization-actions/"
+                                    "python/pip-install.sh"),
+            }
+        ],
```
Comment (Symmetries) [inline data-science-onramp/data-cleaning/clean_test.py]: We don't need initialization_actions anymore

File: data-science-onramp/data-cleaning/clean.sh [hunk 33]
```diff
@@ -0,0 +1,12 @@
+# Submit a PySpark job via the Cloud Dataproc Jobs API
+# Requires having PROJECT_ID, CLUSTER_NAME and BUCKET_NAME set as 
+# environment variables
+
+export CLUSTER_NAME=data-cleaning
+export PROJECT_ID=data-science-onramp
+export BUCKET_NAME=citibikevd
```
Comment (Symmetries) [inline data-science-onramp/data-cleaning/clean.sh:7]: Actually for cleaning I think we should have the user actually type the gcloud command themselves (or use the Google Cloud Console). For the setup we make a bash script only because we wanted to do things behind the scene to prepare the data for the tutorial.

File: data-science-onramp/data-cleaning/clean.py [hunk 34]
```diff
@@ -123,69 +123,61 @@ def compute_end_udf(duration, start, end):
...
-        'end_station_longitude': (angle_udf, FloatType())
+        'start_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'end_station_name': UserDefinedFunction(station_name_udf, StringType()),
+        'tripduration': UserDefinedFunction(trip_duration_udf, IntegerType()),
+        'usertype': UserDefinedFunction(user_type_udf, StringType()),
+        'gender': UserDefinedFunction(gender_udf, StringType()),
+        'start_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'start_station_longitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_latitude': UserDefinedFunction(angle_udf, FloatType()),
+        'end_station_longitude': UserDefinedFunction(angle_udf, FloatType())
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: What is the point of adding UserDefinedFunction on every line here instead of just once below?

File: data-science-onramp/data-cleaning/clean.py [hunk 35]
```diff
@@ -123,69 +123,61 @@ def compute_end_udf(duration, start, end):
...
     df.sample(False, 0.001).show(n=100)
 
-if __name__ == '__main__':
-    main()
+    # Write results to GCS
+    if '--test' in sys.argv:
+        print('Data will not be uploaded to GCS')
+    else:
+        path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+        df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path, mode='overwrite')
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean.py]: Is it considered good practice to overwrite a file instead of printing out an error that it already exists?

File: data-science-onramp/data-cleaning/clean_test.py [hunk 36]
```diff
@@ -0,0 +1,158 @@
...
+    blob = get_blob_from_path(output_location)
+    out = blob.download_as_string().decode("utf-8")
+
+    # tripDuration
+    assert not is_in_table("(\\d+(?:\\.\\d+)?) s", out)
+    assert not is_in_table("(\\d+(?:\\.\\d+)?) min", out)
+    assert not is_in_table("(\\d+(?:\\.\\d+)?) h", out)
+
+    # station latitude & longitude
+    assert not is_in_table("[0-9]+" + u"\u00B0" + "[0-9]+\'[0-9]+\"", out)
```
Comment (vuppalli) [inline data-science-onramp/data-cleaning/clean_test.py]: Maybe we should add an assert that there are floating point values for longitude and latitude?

File: data-science-onramp/data-cleaning/clean.sh (hunk 33 above)
Comment (vuppalli) [reply data-science-onramp/data-cleaning/clean.sh:7]: ^Agreed! We do not need this script at all

File: data-science-onramp/data-cleaning/clean.py (hunk 35 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: Probably not, but I figured for the scope of the tutorial that it wouldn't be an issue. What do you think?

File: data-science-onramp/data-cleaning/clean.py (hunk 34 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: What's the point of storing the initialization parameters if its just going to be used to create a UserDefinedFunction? 

This also makes it much cleaner for the multi-parameter UDFs, specifically bc I'm storing the UDF and its parameters in a dictionary. Also this seems a bit easier to understand in the tutorial.

Thoughts?

File: data-science-onramp/data-cleaning/clean.sh (hunk 22 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.sh:11]: This flag sets the logging level of the job. The special root package controls the root logger level. It is set to FATAL, which is described as "severe errors that cause premature termination."

I think a lower logging level (like WARN for instance) might perhaps produce too much verbose output, especially for the scope of the tutorial. What do you think?

File: data-science-onramp/data-cleaning/clean_test.py (hunk 27 above)
Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean_test.py]: I think it's fine to leave it as-is.

File: data-science-onramp/data-cleaning/clean_test.py (hunk 28 above)
Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean_test.py]: Same as above, I think it's ok.

File: data-science-onramp/data-cleaning/clean_test.py (hunk 29 above)
Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean_test.py]: This, on the other hand, could be moved out. Though it won't be changed often, this would be good to have a bit more visible.

File: data-science-onramp/data-cleaning/clean_test.py (hunk 30 above)
Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean_test.py]: It's not needed to run this specific subset of code, but I believe it is configured on the main cluster that the user will be using to submit all of their Dataproc work. This test, in my opinion, should match that identically (even if it doesn't make sense in isloation) but I defer to your opinion.

File: data-science-onramp/data-cleaning/clean_test.py (hunk 31 above)
Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean_test.py:82]: Dataproc only gives you the URI for the output logs and GCS client libs don't support URI access. You need to parse out the bucket and blob (of which GCS has no native function to do this).

File: data-science-onramp/data-cleaning/clean_test.py (hunk 26 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean_test.py]: I believe this is for a dataproc cluster and I have updated it as such. Please correct me if I'm wrong @Symmetries

File: data-science-onramp/data-cleaning/clean.py (hunk 35 above)
Comment (bradmiro) [inline data-science-onramp/data-cleaning/clean.py]: @TusharK54 A couple things here:

1. Can you add `.coalesce(1)` such that you have `df.coalesce(1).write...`? This will condense your output into a single file. You may also want to split this up over multiple lines to avoid the line-length issue:
```
(
    df
    .coalesce(1)
    .write
    ...
)
```

2. Can you please add logic such that this first writes out to a tmp file and then is copied over to a more conventional location? Feel free to reuse the logic [here](https://github.com/GoogleCloudDataproc/cloud-dataproc/blob/master/codelabs/spark-bigquery/backfill.py#L97)

Comment (bradmiro) [reply data-science-onramp/data-cleaning/clean.py]: Update: Can you and @vuppalli work together to decide which approach makes sense (my suggestion, as is or something entirely different)? Since this will inevitably impact both of your projects, it'l be good for you to be on the same page.

File: data-science-onramp/data-cleaning/clean_test.py (hunk 36 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean_test.py]: good idea

File: data-science-onramp/data-cleaning/clean.py (hunk 35 above)
Comment (vuppalli) [reply data-science-onramp/data-cleaning/clean.py]: Honestly not sure. Why do we even need to add the overwrite?

Comment (vuppalli) [reply data-science-onramp/data-cleaning/clean.py]: I think that adding `.coalesce(1)` might make the script run longer. Since the script already takes a while to complete, and if adding this will make it longer, this might not be a good idea. What do you think? @TusharK54

Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: Otherwise it throws errors if run more than once

Comment (tk755) [reply data-science-onramp/data-cleaning/clean.py]: @bradmiro to address your comments:
1. I just tested it and it seems @vuppalli is right. Using coalesce results in the job taking twice as long as without. Is that a deal breaker?
2. I will go ahead and do that. But could you let me know why we do it like that?

File: data-science-onramp/data-cleaning/clean.sh (hunk 21 above)
Comment (tk755) [reply data-science-onramp/data-cleaning/clean.sh:3]: This should have been done in setup :)

**Comment by bradmiro** [comment]
As long as the test passes, LGTM

**Review by leahecole — APPROVED** [review]
